dunesim/GapChargeTransport/GapChargeTransport_tool.cc

```diff
@@ -0,0 +1,243 @@
+              if (left < x) {
+                right = bbox.MinX();
+                type = GapInfo::GapType::X;
+                if (x > (left + right) / 2.) direction = GapInfo::MovingDirection::RIGHT;
+              }
+            }
+          }
+        }
+      }
+    }
+    std::cout << "Type: " << GapInfo::GapToString(type) << ", left = " << left
+              << ", right = " << right << ", direction = " << GapInfo::GapToString(direction)
+              << std::endl;
+    std::cout.flush();
+    double size = right - left;
+    GapInfo the_gap = GapInfo(type, size, left, right, direction);
+    return the_gap;
+  }
+
+  //in case there will be several modeles, not only linear
+  double GapChargeTransport::ComputeShiftProbability(GapInfo the_gap,
+                                                     double x,
+                                                     double y,
+                                                     double z) const
+  {
+    if (functionType == "linear") { return LinearShiftProbability(the_gap, x, y, z); }
+    else {
+      std::cout << "Unknown function type: " << functionType << ". Defaulting to linear."
+                << std::endl;
+      std::cout.flush();
+      return LinearShiftProbability(the_gap, x, y, z);
+    }
+  }
+
+  //self-explanatory
+  double GapChargeTransport::DistanceToNearestActiveVolume(GapInfo the_gap, double x) const
+  {
+    double distance;
+    if (the_gap.GetDirection() == GapInfo::MovingDirection::LEFT) {
+      distance = x - the_gap.GetLeftBorder();
+    }
+    else {
+      distance = the_gap.GetRightBorder() - x;
+    }
+    std::cout << "left" << the_gap.GetLeftBorder() << " " << x << " " << the_gap.GetRightBorder()
+              << std::endl;
+    std::cout.flush();
+    return distance;
+  }
+
+  //computing the probability to move the charge
+  double GapChargeTransport::LinearShiftProbability(GapInfo the_gap,
+                                                    double x,
+                                                    double y,
+                                                    double z) const
+  {
+
+    double coordinate;
+    switch (the_gap.GetType()) {
+    case GapInfo::GapType::X: coordinate = x; break;
+    case GapInfo::GapType::Y: coordinate = y; break;
+    case GapInfo::GapType::Z: coordinate = z; break;
+    default: coordinate = std::numeric_limits<double>::lowest();
+    }
+    std::cout << GapInfo::GapToString(the_gap.GetType()) << ", coordinate = " << coordinate
+              << std::endl;
+    std::cout.flush();
+    double gapWidth = the_gap.GetSize();
+    double distance = DistanceToNearestActiveVolume(the_gap, coordinate);
+    std::cout << "GapWidth = " << gapWidth << ", distance = " << distance << ", p2 (prob) = " << p2
+              << ", p1 (dist) =  " << p1 << std::endl;
+    std::cout.flush();
+    double probability = p2 - distance / p1 * p2;
+    probability = std::clamp(probability, 0.0, 1.0);
```

Comment:
I was not aware of std::clamp, thank you!

Are there cases where the probability lands outside of 0?  Is it meaningful when it does happen?